hub/util/merge.py

```diff
@@ -0,0 +1,321 @@
+        lca_node,
+        conflict_resolution,
+    )
+    copy_new_tensors(dataset, target_ds, new_tensors)
+
+    if delete_removed_tensors:
+        delete_tensors(dataset, target_deleted_tensors)
+
+    original_node.merge_from(target_node)
+    commit(dataset, f"Merge {target_id} into {dataset.branch}")
+
+
+def get_lca_tensors(ds, lca_id: str):
+    original_id = ds.pending_commit_id
+    ds.checkout(lca_id)
+    lca_tensors: Set[str] = set(ds.tensors.keys())
+    ds.checkout(original_id)
+    return lca_tensors
+
+
+def auto_commit_target_commit(dataset, target_commit_id: str):
+    original_id = dataset.pending_commit_id
+    original_branch = dataset.branch
+    dataset.checkout(target_commit_id)
+    auto_commit(dataset, f"Auto commit before merging into {original_branch}")
+    target_commit_id = dataset.pending_commit_id
+    dataset.checkout(original_id)
+    return target_commit_id
+
+
+def get_changes_commit_ids_for_node(
+    dataset, tensor_name: str, commit_node: Optional[CommitNode], lca_node: CommitNode
+):
+    changes_commit_map = defaultdict(list)
+    current_node = commit_node
+    while current_node and current_node.commit_id != lca_node.commit_id:
+        commit_id = current_node.commit_id
+        if current_node.is_merge_node:
+            changes = get_changes_commit_ids_for_node(
+                dataset, tensor_name, current_node.merge_parent, lca_node
+            )
+            for idx in changes:
+                changes_commit_map[idx].extend(changes[idx])
+        else:
+            try:
+                diff_key = get_tensor_commit_diff_key(tensor_name, commit_id)
+                diff: CommitDiff = dataset.storage.get_hub_object(diff_key, CommitDiff)
+                data_updated = sorted(diff.data_updated)
+                id_tensor_name = get_sample_id_tensor_key(tensor_name)
+                id_tensor = dataset[id_tensor_name]
+                for idx in data_updated:
+                    sample_id = id_tensor[idx].numpy()[0]
+                    changes_commit_map[sample_id].append(commit_id)
+            except KeyError:
+                pass
+        current_node = current_node.parent
+    return changes_commit_map
+
+
+def delete_tensors(dataset, tensor_names: Set[str]):
+    for tensor_name in tensor_names:
+        dataset.delete_tensor(tensor_name)
+
+
+def copy_new_tensors(dataset, target_dataset, tensor_names: Set[str]):
+    for tensor_name in tensor_names:
+        target_tensor = target_dataset[tensor_name]
+        htype = target_tensor.meta.htype
+        sample_compression = target_tensor.meta.sample_compression
+        chunk_compression = target_tensor.meta.chunk_compression
+        dataset.create_tensor(
+            tensor_name,
+            htype=htype,
+            sample_compression=sample_compression,
+            chunk_compression=chunk_compression,
+        )
+        new_tensor = dataset[tensor_name]
+        for item in target_tensor:
+            new_tensor.append(item)
```

Comment:
shouldn't you do chunkwise copy?

Reply:
this will be a problem with version control, not all chunks would be present in the same folder and there would be differences in chunks across versions. While copying here, we want all chunks in same folder.

Reply:
hmm I see, not a priority, we can optimize this later